File: Linked List, Hash/leetcode/203_remove_linked_list_elements/203_remove_linked_list_elements.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

// Function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

// Function to remove elements with a given value from the linked list
struct ListNode* removeElements(struct ListNode* head, int val) {
    struct ListNode* curr = head;
    struct ListNode* prev = NULL;

    // Loop through each node
    while (curr != NULL) {
        if (curr->val == val) {
            if (curr == head) {
                head = curr->next;  // Move head if it's the node to remove
            } else {
                prev->next = curr->next;  // Bypass the node to delete it
            }
            struct ListNode* temp = curr;
            curr = curr->next;
            free(temp);  // Free the memory of the removed node
        } else {
            prev = curr;
            curr = curr->next;
        }
    }
    return head;
}





//////////////////////////////////////////////////////////////////////////////////////////////////////
// Helper function to create a linked list from an array
struct ListNode* createLinkedList(int* values, int size) {
    if (size == 0) return NULL;
    struct ListNode* head = createNode(values[0]);
    struct ListNode* current = head;
    for (int i = 1; i < size; i++) {
        current->next = createNode(values[i]);
        current = current->next;
    }
    return head;
}

// Helper function to print the linked list
void printLinkedList(struct ListNode* head) {
    struct ListNode* current = head;
    while (current != NULL) {
        printf("%d", current->val);
        if (current->next != NULL) printf(" -> ");
        current = current->next;
    }
    printf("\n");
}

// Function to free the linked list
void freeLinkedList(struct ListNode* head) {
    while (head != NULL) {
        struct ListNode* temp = head;
        head = head->next;
        free(temp);
    }
}

int main() {
    // Example usage
    int values[] = {1, 2, 6, 3, 4, 5, 6};
    int size = sizeof(values) / sizeof(values[0]);

    struct ListNode* head = createLinkedList(values, size);

    printf("Original list:\n");
    printLinkedList(head);

    // Remove elements with value 6
    head = removeElements(head, 6);

    printf("List after removing elements with value 6:\n");
    printLinkedList(head);

    // Free the list
    freeLinkedList(head);

    return 0;
}

File: Linked List, Hash/leetcode/203_remove_linked_list_elements/203_remove_linked_list_elements.py
'''
head of liked list가 주어지고, val이 주어지면, liked list에서 val 값을 갖는 모든 노드는 지우고, 
head를 반환

linked list에서 searching, deleting을 하면 됨.


Example 2:

Input: head = [], val = 1
Output: []


Example 3:

Input: head = [7,7,7,7], val = 7
Output: []
'''

from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        '''
        Optional[ListNode]
        head : ListNode class로 주어주든, 혹은 None으로 주어진다는 뜻.
        '''
        
        if not head:
            return None
        
  
        current = head
        prev = None

        while(current):
            
            #find node and pop
            if current.val == val:
                if current == head:
                    head = current.next
                
                else:
                    prev.next = current.next

            else:
                prev = current #넘어가는게 current == val이 아닐 때만, 맞으면 넘어가면 안됨.
            current = current.next

        return head
    

#번외
def create_liked_list(values : list):
    if not values:
        return None
    
    head = ListNode(values[0])
    current = head
    for value in values[1:]:
        current.next = ListNode(value)
        current = current.next
    return head



def print_liked_list(head : Optional[ListNode]):
    current = head
    elements =[]
    while(current):
        elements.append(current.val)
        current = current.next

    print(" -> ".join(map(str, elements)))
        

def main():
    
    values = [1,2,6,3,4,5,6]
    values =[7,7,7,7]
    values = [1,2,2,1]
    head = create_liked_list(values)
    print_liked_list(head)
    val = 2

    sol = Solution()
    new_head = sol.removeElements(head, val)
    print_liked_list(new_head)
    

if __name__ == "__main__":
    main()